src/Symfony/Component/Serializer/Normalizer/ChainNormalizer.php

```diff
@@ -0,0 +1,213 @@
+     */
+    public function __construct(array $normalizers = [])
+    {
+        foreach ($normalizers as $normalizer) {
+            $this->addNormalizer($normalizer);
+        }
+    }
+
+    /**
+     * Add Normalizer last in the line.
+     */
+    public function addNormalizer(NormalizerInterface $normalizer): void
+    {
+        if ($normalizer instanceof NormalizerAwareInterface) {
+            $normalizer->setNormalizer($this);
+        }
+        if (null !== $this->serializer && $normalizer instanceof SerializerAwareInterface) {
+            $normalizer->setSerializer($this->serializer);
+        }
+
+        $this->normalizers[] = $normalizer;
+        $this->normalizerCache = [];
+        $this->supportedCache = [];
+    }
+
+    public function normalize(mixed $object, ?string $format = null, array $context = []): array|string|int|float|bool|\ArrayObject|null
+    {
+        // If a normalizer supports the given data, use it
+        if ($normalizer = $this->getNormalizer($object, $format, $context)) {
+            return $normalizer->normalize($object, $format, $context);
+        }
+
+        if (null === $object || \is_scalar($object)) {
+            return $object;
+        }
+
+        if (\is_array($object) && !$object && ($context[Serializer::EMPTY_ARRAY_AS_OBJECT] ?? false)) {
+            return new \ArrayObject();
+        }
+
+        if (is_iterable($object)) {
+            if ($object instanceof \Countable && ($context[AbstractObjectNormalizer::PRESERVE_EMPTY_OBJECTS] ?? false) && !\count($object)) {
+                return new \ArrayObject();
+            }
+
+            $normalized = [];
+            foreach ($object as $key => $val) {
+                $normalized[$key] = $this->normalize($val, $format, $context);
+            }
+
+            return $normalized;
+        }
+
+        if (\is_object($object)) {
+            if ([] === $this->normalizers) {
+                throw new LogicException('You must register at least one normalizer to be able to normalize objects.');
+            }
+
+            throw new NotNormalizableValueException(sprintf('Could not normalize object of type "%s", no supporting normalizer found.', get_debug_type($object)));
+        }
+
+        throw new NotNormalizableValueException('An unexpected value could not be normalized: '.(!\is_resource($object) ? var_export($object, true) : sprintf('"%s" resource', get_resource_type($object))));
+    }
+
+    public function supportsNormalization(mixed $data, ?string $format = null, array $context = []): bool
+    {
+        return null !== $this->getNormalizer($data, $format, $context);
+    }
+
+    public function getSupportedTypes(?string $format): array
+    {
+        if (null === $format) {
+            $format = '__null__';
+        }
+
+        if (!isset($this->supportedCache[$format])) {
+            foreach ($this->normalizers as $normalizer) {
+                $this->supportedCache + $normalizer->getSupportedTypes($format);
```

Comment:
this logic is flawed:

- it does not write in `$this->supportedCache` at all, so if it is not set, you will get a notice below because it still won't be set
- it does not report that this normalizer knows how to normalize any non-object value without the help of any registered normalizer
- it does not properly deal with cases where multiple normalizers report support for the same type with different cacheability status (for which even the *order* of answers can be important, see the complexity in `getSupportedTypes`)

IMO, the only sane implementation of that method (especially considering that the only case where it will be used is when nesting a ChainNormalizer inside another so we won't need it for 99.999% of use cases) is `return ['*' => false];`

Reply:
Note that the same is true for the Denormalizer

Reply:
Thank you. I've updated this and added a test to prove correctness.